### PR TITLE
bug fix to EG PF isolation map

### PR DESCRIPTION
**Pull request by Sam-Harper**
Dear All,

@matteosan1 @lgray 

This change effects the product named particleBasedIsolation. It is used by E/gamma in the computation of electron and photon particle flow isolation at analysis level. It is a final product of the reconstruction chain, nothing in RECO depends on it. 

This map keeps track of which particle flow candidates make up the GedGsfElectron/GedPhoton. In the case the GedEle/Pho pass PF selection, this should only be the identified particle flow ele/pho. In the case the ele/pho fails PF selection, the ele/pho is broken up into charged/neutral/photons as PFAlgo sees fit.  The idea of the map is to flag which particles came from this break up and then veto them in the isolation sum. If you fail to do this, you will get a nice peak of PFIsol/et at 1 as all the ele/phos energy is dumped into the isolation region.

We have seen that the map is slightly bugged and is as such slightly greedy and sums in objects it shouldnt, particularly in the case of photon isolation. After extensive testing, we have decided that the map is to stay the same for neutral and charged hadrons but we will require the PFPhoton's leading ECAL element to be either the GedEle/Pho's supercluster or one of its subclusters in addition to previous requirements. This change fixes this, it is a small improvement but an improvement none the less.

https://indico.cern.ch/event/370494/contribution/3
https://indico.cern.ch/event/305124/contribution/4
https://sharper.web.cern.ch/sharper/cms/talks/debugingEGIsoMap_160215_update.pdf

Cheers,
Sam

### CODE REVIEW

**Comment by slava77** [comment]
Hi Sam,

Please change the file name to PF*
It's a more common use case in CMSSW

**Comment by Sam-Harper** [comment]
Done. Just re-running the tests now and will push it shortly.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
Starting matrix tests

File: RecoEgamma/EgammaIsolationAlgos/src/PFBlockBasedIsolation.cc [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  const reco::PFBlockRef egblock = ieg->first;
+
+
+  unsigned nObj = pfCandidateHandle->size();
+  for(unsigned int lCand=0; lCand < nObj; lCand++) {
+
+    reco::PFCandidateRef pfCandRef(reco::PFCandidateRef(pfCandidateHandle,lCand));
+
+    float dR = 0.0;
+    if( coneSize_ < 10.0 ) {
```
Comment (cvuosalo) [inline RecoEgamma/EgammaIsolationAlgos/src/PFBlockBasedIsolation.cc]: This 10.0 looks like a magic number, which CMS coding rules frown upon. Could you replace it with an appropriately named const float?

**Comment by cvuosalo** [comment]
@degano, @nclopezo: Yet another Jenkins failure -- "Comparison with the baseline" stuck in Queued state. I don't actually need the results because I ran my own matrix tests, but it is worrisome that in the last week there have been several Jenkins failures.

**Comment by cvuosalo** [comment]
+1

For #7922 3e3cb6ba267f2796551d42441c9c721b7c873613

This PR is a bug fix to the E/gamma isolation map. It is not expected to change reconstruction because it involves isolation calculation that occurs after reconstruction.

The code change is satisfactory, as are the Jenkins test results. Matrix tests against baseline CMSSW_7_4_X_2015-02-24-0200 show no significant differences.

**Comment by davidlange6** [comment]
+1

**Pull request merged**